armory/docker/images.py

```diff
@@ -32,3 +36,218 @@
+    log.info(f"Retrieved canonical image name for {image_name} as {canon_image_name}")
+
+    prev_release = last_armory_release(canon_image_name)
+    if canon_image_name != prev_release:  # currently on hashed dev branch
+        log.trace(f"asking local docker for image {canon_image_name}")
+        try:
+            docker_client.images.get(canon_image_name)
+            log.success(f"found docker image {image_name} as {canon_image_name}")
+            return canon_image_name
+        except docker.errors.ImageNotFound:
+            log.trace(f"image {canon_image_name} not found")
+            log.info(f"reverting to previous release tag image {prev_release}")
+        except requests.exceptions.HTTPError:
+            log.trace(f"http error when looking for image {canon_image_name}")
+            raise
+
+    log.trace(f"asking local docker for image {prev_release}")
+    try:
+        docker_client.images.get(prev_release)
+        log.success(f"found docker image {image_name} as {prev_release}")
+        return prev_release
+    except docker.errors.ImageNotFound:
+        log.trace(f"image {prev_release} not found")
+    except requests.exceptions.HTTPError:
+        log.trace(f"http error when looking for image {prev_release}")
+        raise
+
+    log.info(f"image {prev_release} not found. downloading...")
+    try:
+        pull_verbose(docker_client, prev_release)
+        return prev_release
+    except docker.errors.NotFound:
+        log.error(f"Image {prev_release} could not be downloaded")
+        raise ValueError(
+            "You are attempting to pull an unpublished armory docker image.\n"
+            "This is likely because you're running armory from a dev branch. "
+            "If you want a stable release with "
+            "published docker images try pip installing 'armory-testbed' "
+            "or using out one of the release branches on the git repository. "
+            "If you'd like to continue working on the developer image please "
+            "build it from source on your machine as described here:\n"
+            "https://armory.readthedocs.io/en/latest/contributing/#development-docker-containers\n"
+            "python docker/build.py --help"
+        )
+    except requests.exceptions.ConnectionError:
+        log.error("Docker connection refused. Is Docker Daemon running?")
+        raise
+
+
+def pull_verbose(docker_client, repository, tag=None):
+    """
+    Use low-level docker-py API to show status while pulling docker containers.
+        Attempts to replicate docker command line output if we are showing progress.
+    """
+    if tag is None and ":" in repository:
+        repository, tag = repository.split(":")
+    elif ":" in repository:
+        raise ValueError(
+            f"cannot set tag kwarg and have tag in repository arg {repository}"
+        )
+    elif tag is None:
+        log.info("empty tag is set to latest by API")
+        tag = "latest"
+
+    if not is_progress():
```

Comment:
thanks for preserving `not is_progress`; I wondered if that was a Good Thing or just a good thing to me.